pkg/terraformer/terraformer.go

```diff
@@ -205,3 +229,93 @@ func (t *Terraformer) executeTerraform(ctx context.Context, command Command) err
 	log.Info("terraform process finished successfully", "command", command)
 	return nil
 }
+
+func (t *Terraformer) addFinalizer(ctx context.Context) error {
+	logger := t.stepLogger("add-finalizer")
+	return t.updateObjects(ctx, logger, controllerutil.AddFinalizer)
+
+}
+
+func (t *Terraformer) removeFinalizer(ctx context.Context) error {
+	logger := t.stepLogger("remove-finalizer")
+	return t.updateObjects(ctx, logger, controllerutil.RemoveFinalizer)
+}
+
+func (t *Terraformer) terraformObjects() []controllerutil.Object {
+	return []controllerutil.Object{
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.VariablesSecretName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.ConfigurationConfigMapName,
+			},
+		},
+		&corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: t.config.Namespace,
+				Name:      t.config.StateConfigMapName,
+			},
+		},
+	}
+}
+
+func (t *Terraformer) updateObjects(ctx context.Context, log logr.Logger, patchObj func(controllerutil.Object, string)) error {
+	allErrors := &multierror.Error{
+		ErrorFormat: utils.NewErrorFormatFuncWithPrefix("failed to update object finalizer"),
+	}
+
+	log.Info("updating finalizers for terraform resources")
+	for _, obj := range t.terraformObjects() {
+		if err := t.updateObjectFinalizers(ctx, log, obj, patchObj); err != nil {
+			allErrors = multierror.Append(allErrors, err)
+		}
+	}
+
+	err := allErrors.ErrorOrNil()
+	if err != nil {
+		log.Error(err, "failed to updated finalizers for all terraform resources")
+	} else {
+		log.Info("successfully updated finalizers for terraform resources")
+	}
+	return err
+}
+
+func (t *Terraformer) updateObjectFinalizers(ctx context.Context, log logr.Logger, obj controllerutil.Object, patchObj func(controllerutil.Object, string)) error {
+	key, err := client.ObjectKeyFromObject(obj)
+	if err != nil {
+		log.Error(err, "failed to construct key", "object", obj)
+		return err
+	}
+
+	for i := 0; i < maxPatchRetries; i++ {
+		err = t.client.Get(ctx, key, obj)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.V(1).Info("create empty object", "key", key)
+				patchObj(obj, TerraformerFinalizer)
+				return t.client.Create(ctx, obj)
```

Comment:
Can you add a small test / check to some existing test (e.g. the pod e2e test), that verifies this behaviour (that the finalizer is added to the state configMap if it didn't exist before)?

Reply:
Done.